src/sagemaker_server.cc

```diff
@@ -138,4 +283,285 @@ SagemakerAPIServer::Create(
+  evhtp_send_reply(req, EVHTP_RES_OK);
+}
+
+void
+SagemakerAPIServer::SageMakerMMELoadModel(
+    evhtp_request_t* req,
+    const std::unordered_map<std::string, std::string> parse_map)
+{
+  std::string repo_path = parse_map.at("url");
+  std::string model_name = parse_map.at("model_name");
+
+  /* Error out if there's more than one subdir/version within
+   * supplied model repo, as ensemble in MME is not (currently)
+   * supported
+   */
+  DIR* dir;
+  struct dirent* ent;
+  int dir_count = 0;
+  if ((dir = opendir(repo_path.c_str())) != NULL) {
+    while ((ent = readdir(dir)) != NULL) {
+      if ((ent->d_type == DT_DIR) && (strcmp(ent->d_name, ".") == 0) &&
+          (strcmp(ent->d_name, "..") == 0)) {
+        dir_count += 1;
+      }
+      if (dir_count > 1) {
+        HTTP_RESPOND_IF_ERR(
+            req,
+            TRITONSERVER_ErrorNew(
+                TRITONSERVER_ERROR_INTERNAL,
+                "More than one version or model directories found. Note that "
+                "Ensemble models are not supported in SageMaker MME mode."));
+        closedir(dir);
+        return;
+      }
+    }
+    closedir(dir);
+  }
+
+  std::vector<const TRITONSERVER_Parameter*> subdir_modelname_map;
+
+  /* Split repo path into three parts:
+   * /opt/ml/models/<hash>/model/optional_customer_subdir
+   * 1st repo_parent_path: /opt/ml/models/<hash>
+   * 2nd subdir: model
+   * 3rd customer_subdir: optional_customer_subdir
+   */
+
+  std::string repo_parent_path, subdir, customer_subdir;
+  RE2::FullMatch(
+      repo_path, model_path_regex_, &repo_parent_path, &subdir,
+      &customer_subdir);
+
+  auto param = TRITONSERVER_ParameterNew(
+      subdir.c_str(), TRITONSERVER_PARAMETER_STRING, model_name.c_str());
+
+  if (param != nullptr) {
+    subdir_modelname_map.emplace_back(param);
+  } else {
+    HTTP_RESPOND_IF_ERR(
+        req, TRITONSERVER_ErrorNew(
+                 TRITONSERVER_ERROR_INTERNAL,
+                 "unexpected error on creating Triton parameter"));
+  }
+
+  /* Register repository with model mapping */
+  TRITONSERVER_Error* err = nullptr;
+  err = TRITONSERVER_ServerRegisterModelRepository(
+      server_.get(), repo_parent_path.c_str(), subdir_modelname_map.data(),
+      subdir_modelname_map.size());
+
+  // If a model_name is reused i.e. model_name is already mapped, return a 409
+  if ((err != nullptr) &&
+      (TRITONSERVER_ErrorCode(err) == TRITONSERVER_ERROR_ALREADY_EXISTS)) {
+    EVBufferAddErrorJson(req->buffer_out, err);
+    evhtp_send_reply(req, EVHTP_RES_CONFLICT); /* 409 */
+    TRITONSERVER_ErrorDelete(err);
+    return;
+  } else if (err != nullptr) {
+    EVBufferAddErrorJson(req->buffer_out, err);
+    evhtp_send_reply(req, EVHTP_RES_BADREQ);
+    TRITONSERVER_ErrorDelete(err);
+    return;
+  }
+
+  /* TODO: Check if we should use TRITONSERVER_ServerLoadModelWithParameters */
+  TRITONSERVER_Error* load_err =
+      nullptr; /* TODO: Check if we can re-use previous object*/
```

Comment:
Yes, `err` can be re-used since it will still be `nullptr` if control reaches this point